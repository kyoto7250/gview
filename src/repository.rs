use git2::{Commit, ObjectType, Oid, Repository, TreeWalkMode, TreeWalkResult};
use std::{
    io::{BufRead, BufReader},
    path::{Path, PathBuf},
};

const MAX_FILE_SIZE: usize = 16 * 1024; // 16KB

#[derive(Debug)]
pub struct CommitRow {
    pub _author: String,
    pub commit: Oid,
    pub number: usize,
    pub line: String,
}

impl CommitRow {
    pub fn new(author: String, commit: Oid, number: usize, line: String) -> CommitRow {
        Self {
            _author: author,
            commit,
            number,
            line,
        }
    }
}

pub struct RepositoryInfo {
    repository: Repository,
    oid: Oid,
}

impl std::fmt::Debug for RepositoryInfo {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("RepositoryInfo")
            .field("oid", &self.oid)
            .finish()
    }
}

impl RepositoryInfo {
    pub fn new() -> anyhow::Result<Self> {
        let repo_path = std::env::current_dir()?;
        let repository = Repository::discover(repo_path)?;
        let oid = repository.head()?.target().unwrap();
        Ok(Self { repository, oid })
    }

    // NOTE: this function should only be used during testing.
    pub fn _from_parts(repository: Repository, oid: Oid) -> Self {
        Self { repository, oid }
    }

    pub fn current_commit(&mut self) -> anyhow::Result<(String, String)> {
        let commit = self.repository.find_commit(self.oid)?;
        let commit_message = commit.message().unwrap_or("No commit message");
        Ok((self.oid.to_string(), commit_message.to_owned()))
    }

    pub fn set_parent_commit(&mut self) {
        let commit = self.repository.find_commit(self.oid).unwrap();
        if commit.parent_count() > 0 {
            self.oid = commit.parent(0).unwrap().id();
        }
    }

    pub fn set_next_commit(&mut self) -> anyhow::Result<(String, String)> {
        let next_commit_id = {
            let next_commit = self.find_next_commit()?;
            next_commit.map(|next_commit| next_commit.id())
        };

        if let Some(next_commit_id) = next_commit_id {
            self.oid = next_commit_id;
        }
        self.current_commit()
    }

    fn find_next_commit(&mut self) -> anyhow::Result<Option<Commit>> {
        let commit = self.repository.find_commit(self.oid)?;
        let mut revwalk = self.repository.revwalk()?;
        revwalk.push_head()?;
        revwalk.set_sorting(git2::Sort::REVERSE)?;

        let mut next_commit_found = false;
        for oid_result in revwalk {
            let oid = oid_result?;
            let rev_commit = self.repository.find_commit(oid)?;

            if next_commit_found {
                return Ok(Some(rev_commit));
            }

            if rev_commit.id() == commit.id() {
                next_commit_found = true;
            }
        }

        Ok(None)
    }

    pub fn get_content(&mut self, filename: String) -> anyhow::Result<Vec<CommitRow>> {
        if filename == *"not found" {
            return Ok(vec![]);
        }
        let path = Path::new(&filename);
        let blame = self.repository.blame_file(path, None)?;
        let commit = self.repository.head()?.peel_to_commit()?;
        let tree = commit.tree()?;
        let blob = tree
            .get_path(path)?
            .to_object(&self.repository)?
            .peel_to_blob()?;
        let reader = BufReader::new(blob.content());
        let mut content = vec![];
        for (i, line) in reader.lines().enumerate() {
            if let (Ok(line), Some(hunk)) = (line, blame.get_line(i + 1)) {
                let signature = hunk.orig_signature();
                let author = signature.name().unwrap_or("Unknown");
                let commit_id = hunk.final_commit_id();
                // let line_number = hunk.final_start_line();
                let row = CommitRow::new(author.to_owned(), commit_id, i + 1, line);
                content.push(row);
            }
        }

        Ok(content)
    }
    pub fn get_commit_history(&self) -> anyhow::Result<Vec<(String, String)>> {
        let mut revwalk = self.repository.revwalk()?;
        revwalk.push_head()?;
        revwalk.set_sorting(git2::Sort::TIME)?;

        let mut commits = Vec::new();
        for oid_result in revwalk {
            let oid = oid_result?;
            let commit = self.repository.find_commit(oid)?;
            let commit_message = commit
                .message()
                .unwrap_or("No commit message")
                .lines()
                .next()
                .unwrap_or("")
                .to_string();
            commits.push((oid.to_string(), commit_message));
        }

        Ok(commits)
    }

    pub fn get_current_commit_id(&self) -> String {
        self.oid.to_string()
    }

    pub fn set_commit_by_id(&mut self, commit_id: &str) -> anyhow::Result<()> {
        let oid = if commit_id.len() == 40 {
            // Full commit ID
            git2::Oid::from_str(commit_id)?
        } else {
            // Short commit ID - need to resolve it
            let mut revwalk = self.repository.revwalk()?;
            revwalk.push_head()?;

            let mut found_oid = None;
            let mut match_count = 0;
            for oid_result in revwalk {
                let oid = oid_result?;
                let oid_str = oid.to_string();
                if oid_str.starts_with(commit_id) {
                    found_oid = Some(oid);
                    match_count += 1;
                    if match_count > 1 {
                        return Err(anyhow::anyhow!(
                            "Ambiguous commit ID: multiple commits match '{}'",
                            commit_id
                        ));
                    }
                }
            }

            found_oid.ok_or_else(|| anyhow::anyhow!(format!("Commit '{}' not found", commit_id)))?
        };

        // Verify the commit exists before setting it
        self.repository.find_commit(oid)?;
        self.oid = oid;
        Ok(())
    }

    pub fn recursive_walk(&mut self) -> anyhow::Result<Vec<String>> {
        let head = self.repository.find_commit(self.oid)?;
        let tree = head.tree()?;
        let mut results: Vec<String> = vec![];
        let mut path_stack: Vec<PathBuf> = vec![PathBuf::new()];
        let _ = tree.walk(TreeWalkMode::PreOrder, |root, entry| {
            if let Some(name) = entry.name() {
                let mut current_path = PathBuf::from(root);
                current_path.push(name);

                if let Ok(obj) = entry.to_object(&self.repository) {
                    match obj.kind() {
                        Some(ObjectType::Blob) => {
                            let blob = obj.peel_to_blob().unwrap();
                            let content = blob.content();
                            if content.len() < MAX_FILE_SIZE && content.is_ascii() {
                                results.push(current_path.to_string_lossy().to_string());
                            }
                        }
                        Some(ObjectType::Tree) => {
                            path_stack.push(current_path.clone());
                        }
                        _ => (),
                    }
                }
            }
            TreeWalkResult::Ok
        });

        Ok(results)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::env;
    use std::fs;
    use std::io::Write;

    fn setup_test_repo_with_file() -> (Repository, String) {
        use std::time::{SystemTime, UNIX_EPOCH};
        let timestamp = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_nanos();
        let test_dir = env::temp_dir().join(format!("gview_test_repo_{}", timestamp));
        let _ = fs::remove_dir_all(&test_dir);
        fs::create_dir_all(&test_dir).unwrap();

        let repo = Repository::init(&test_dir).unwrap();

        // Create a test file
        let test_file_path = test_dir.join("test.txt");
        let mut file = fs::File::create(&test_file_path).unwrap();
        file.write_all(b"line 1\nline 2\nline 3\n").unwrap();

        // Add and commit the file
        let mut index = repo.index().unwrap();
        index.add_path(std::path::Path::new("test.txt")).unwrap();
        index.write().unwrap();

        let signature = git2::Signature::new(
            "Test User",
            "test@example.com",
            &git2::Time::new(1234567890, 0),
        )
        .unwrap();
        let tree_id = index.write_tree().unwrap();
        let tree = repo.find_tree(tree_id).unwrap();

        let _ = repo.commit(
            Some("HEAD"),
            &signature,
            &signature,
            "Add test file",
            &tree,
            &[],
        );

        drop(tree);
        (repo, "test.txt".to_string())
    }

    fn setup_empty_repo() -> Repository {
        use std::time::{SystemTime, UNIX_EPOCH};
        let timestamp = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_nanos();
        let test_dir = env::temp_dir().join(format!("gview_empty_repo_{}", timestamp));
        let _ = fs::remove_dir_all(&test_dir);
        fs::create_dir_all(&test_dir).unwrap();

        let repo = Repository::init(&test_dir).unwrap();

        let signature = git2::Signature::new(
            "Test User",
            "test@example.com",
            &git2::Time::new(1234567890, 0),
        )
        .unwrap();
        let tree_id = {
            let mut index = repo.index().unwrap();
            index.write_tree().unwrap()
        };
        let tree = repo.find_tree(tree_id).unwrap();

        let _ = repo.commit(
            Some("HEAD"),
            &signature,
            &signature,
            "Initial commit",
            &tree,
            &[],
        );

        drop(tree);
        repo
    }

    #[test]
    fn test_commit_row_new() {
        let oid = Oid::from_str("0123456789abcdef0123456789abcdef01234567").unwrap();
        let row = CommitRow::new(
            "test_author".to_string(),
            oid,
            42,
            "println!(\"Hello, world!\");".to_string(),
        );

        assert_eq!(row._author, "test_author");
        assert_eq!(row.commit, oid);
        assert_eq!(row.number, 42);
        assert_eq!(row.line, "println!(\"Hello, world!\");");
    }

    #[test]
    fn test_repository_info_current_commit() {
        let repo = setup_empty_repo();
        let head_commit = repo.head().unwrap().target().unwrap();

        let mut repo_info = RepositoryInfo {
            repository: repo,
            oid: head_commit,
        };

        let result = repo_info.current_commit().unwrap();
        assert_eq!(result.0.len(), 40); // SHA length
        assert_eq!(result.1, "Initial commit");
    }

    #[test]
    fn test_set_commit_by_id_full_hash() {
        let (repo, _) = setup_test_repo_with_file();
        let head_commit = repo.head().unwrap().target().unwrap();
        let head_commit_str = head_commit.to_string();

        let mut repo_info = RepositoryInfo {
            repository: repo,
            oid: head_commit,
        };

        // Test setting by full commit ID
        let result = repo_info.set_commit_by_id(&head_commit_str);
        assert!(result.is_ok());
        assert_eq!(repo_info.oid, head_commit);
    }

    #[test]
    fn test_set_commit_by_id_short_hash() {
        let (repo, _) = setup_test_repo_with_file();
        let head_commit = repo.head().unwrap().target().unwrap();
        let head_commit_str = head_commit.to_string();
        let short_commit = &head_commit_str[..7]; // Use 7 characters

        let mut repo_info = RepositoryInfo {
            repository: repo,
            oid: head_commit,
        };

        // Test setting by short commit ID
        let result = repo_info.set_commit_by_id(short_commit);
        assert!(result.is_ok());
        assert_eq!(repo_info.oid, head_commit);
    }

    #[test]
    fn test_set_commit_by_id_invalid() {
        let (repo, _) = setup_test_repo_with_file();
        let head_commit = repo.head().unwrap().target().unwrap();

        let mut repo_info = RepositoryInfo {
            repository: repo,
            oid: head_commit,
        };

        // Test setting by invalid commit ID
        let result = repo_info.set_commit_by_id("invalid123");
        assert!(result.is_err());
        assert_eq!(repo_info.oid, head_commit); // Should remain unchanged
    }

    #[test]
    fn test_repository_info_set_parent_commit() {
        let (repo, _) = setup_test_repo_with_file();
        let _head_commit = repo.head().unwrap().target().unwrap();

        // Create a second commit
        let signature = git2::Signature::new(
            "Test User",
            "test@example.com",
            &git2::Time::new(1234567890, 0),
        )
        .unwrap();
        let tree = repo
            .head()
            .unwrap()
            .peel_to_commit()
            .unwrap()
            .tree()
            .unwrap();
        let parent_commit = repo.head().unwrap().peel_to_commit().unwrap();
        let second_commit_oid = repo
            .commit(
                Some("HEAD"),
                &signature,
                &signature,
                "Second commit",
                &tree,
                &[&parent_commit],
            )
            .unwrap();

        drop(tree);
        drop(parent_commit);

        let mut repo_info = RepositoryInfo {
            repository: repo,
            oid: second_commit_oid,
        };

        let original_oid = repo_info.oid;
        repo_info.set_parent_commit();

        // Should now be pointing to the parent commit
        assert_ne!(original_oid, repo_info.oid);
    }

    #[test]
    fn test_repository_info_set_parent_commit_no_parent() {
        let repo = setup_empty_repo();
        let head_commit = repo.head().unwrap().target().unwrap();

        let mut repo_info = RepositoryInfo {
            repository: repo,
            oid: head_commit,
        };

        let original_oid = repo_info.oid;
        repo_info.set_parent_commit();

        // Should remain the same as it has no parent
        assert_eq!(original_oid, repo_info.oid);
    }

    #[test]
    fn test_get_content_not_found_special_case() {
        let repo = setup_empty_repo();
        let head_commit = repo.head().unwrap().target().unwrap();

        let mut repo_info = RepositoryInfo {
            repository: repo,
            oid: head_commit,
        };

        let result = repo_info.get_content("not found".to_string()).unwrap();
        assert!(result.is_empty());
    }

    #[test]
    fn test_get_content_with_file() {
        let (repo, filename) = setup_test_repo_with_file();
        let head_commit = repo.head().unwrap().target().unwrap();

        let mut repo_info = RepositoryInfo {
            repository: repo,
            oid: head_commit,
        };

        let result = repo_info.get_content(filename).unwrap();
        assert_eq!(result.len(), 3); // 3 lines
        assert_eq!(result[0].line, "line 1");
        assert_eq!(result[1].line, "line 2");
        assert_eq!(result[2].line, "line 3");
        assert_eq!(result[0].number, 1);
        assert_eq!(result[1].number, 2);
        assert_eq!(result[2].number, 3);
    }

    #[test]
    fn test_recursive_walk_empty_repo() {
        let repo = setup_empty_repo();
        let head_commit = repo.head().unwrap().target().unwrap();

        let mut repo_info = RepositoryInfo {
            repository: repo,
            oid: head_commit,
        };

        let result = repo_info.recursive_walk().unwrap();
        assert!(result.is_empty());
    }

    #[test]
    fn test_recursive_walk_with_file() {
        let (repo, _) = setup_test_repo_with_file();
        let head_commit = repo.head().unwrap().target().unwrap();

        let mut repo_info = RepositoryInfo {
            repository: repo,
            oid: head_commit,
        };

        let result = repo_info.recursive_walk().unwrap();
        assert_eq!(result.len(), 1);
        assert_eq!(result[0], "test.txt");
    }

    #[test]
    fn test_find_next_commit_no_next() {
        let repo = setup_empty_repo();
        let head_commit = repo.head().unwrap().target().unwrap();

        let mut repo_info = RepositoryInfo {
            repository: repo,
            oid: head_commit,
        };

        let result = repo_info.find_next_commit().unwrap();
        assert!(result.is_none());
    }

    #[test]
    fn test_set_next_commit_no_next() {
        let repo = setup_empty_repo();
        let head_commit = repo.head().unwrap().target().unwrap();

        let mut repo_info = RepositoryInfo {
            repository: repo,
            oid: head_commit,
        };

        let original_oid = repo_info.oid;
        let result = repo_info.set_next_commit().unwrap();

        // Should remain the same as there's no next commit
        assert_eq!(original_oid, repo_info.oid);
        assert_eq!(result.1, "Initial commit");
    }
}
