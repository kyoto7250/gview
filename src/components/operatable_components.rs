use std::sync::{Arc, Mutex};

use crate::repository::RepositoryInfo;
use crossterm::event::KeyCode;
use ratatui::{layout::Rect, Frame};

// rust enum pass the operation command
pub enum Message {
    MultipleTimes(MultipleTimesOperation),
    Once(OnceOperation),
    NoAction,
    Error { message: String },
}

pub enum MultipleTimesOperation {
    Filtering {
        query: String,
    },
    SetUp {
        repository: Arc<Mutex<RepositoryInfo>>,
    },
    ChangeShowCommit,
}

pub enum OnceOperation {
    ShowFile { file: String },
    JumpToContentView,
    JumpToFiler,
}

#[derive(Clone, Copy, Debug)]
pub enum Focus {
    OFF,
    ON,
}

pub trait OperatableComponent {
    fn draw(&mut self, frame: &mut Frame, rect: Rect);
    fn process_focus(&mut self);
    fn process_events(&mut self, events: KeyCode) -> Message;
    fn handle_message(&mut self, message: &Message) -> Message;
}
