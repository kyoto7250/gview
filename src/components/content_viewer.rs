use std::sync::{Arc, Mutex};

use crossterm::event::KeyCode;
use ratatui::{
    layout::Rect,
    style::{Color, Style, Stylize},
    widgets::{Block, Paragraph, Wrap},
    Frame,
};

use crate::repository::{CommitRow, RepositoryInfo};

use super::operatable_components::{Focus, Message, OnceOperation, OperatableComponent};

pub enum ShowMode {
    WithLine,
    WithBlame,
    NoLine,
}

impl ShowMode {
    fn next(&mut self) -> ShowMode {
        match self {
            Self::WithLine => Self::WithBlame,
            Self::WithBlame => Self::NoLine,
            Self::NoLine => Self::WithLine,
        }
    }

    fn concat(&mut self, rows: Vec<CommitRow>) -> String {
        match self {
            Self::NoLine => rows
                .iter()
                .map(|row| row.line.to_owned())
                .collect::<Vec<String>>()
                .join("\n"),
            Self::WithLine => {
                // TODO: maximum size
                rows.iter()
                    .map(|row| format!("{} | {} ", row.number, row.line.to_owned()))
                    .collect::<Vec<String>>()
                    .join("\n")
            }
            Self::WithBlame => {
                // TODO: maximum size
                rows.iter()
                    .map(|row| format!("{} | {} ", row.commit, row.line.to_owned()))
                    .collect::<Vec<String>>()
                    .join("\n")
            }
        }
    }
}

pub struct ContentViewer {
    focus: Focus,
    title: String,
    content: String,
    context_size: usize,
    scroll_position: usize,
    height: usize,
    repository: Arc<Mutex<RepositoryInfo>>,
    mode: ShowMode,
}

impl ContentViewer {
    pub fn new(repository: Arc<Mutex<RepositoryInfo>>) -> Self {
        Self {
            focus: Focus::Off,
            title: "Content Viewer".to_owned(),
            content: "".to_owned(),
            repository,
            context_size: 0,
            height: 0,
            scroll_position: 0,
            mode: ShowMode::WithLine,
        }
    }

    fn _handle_message(&mut self, message: &Message) -> Message {
        match message {
            Message::Once(OnceOperation::ShowFile { file }) => {
                // update content view
                file.clone_into(&mut self.title);
                let mut repository = self.repository.lock().unwrap();

                if let Ok(rows) = repository.get_content(file.to_owned()) {
                    self.content = self.mode.concat(rows);
                    self.scroll_position = 0
                } else {
                    return Message::Error {
                        _message: "failed to get content".to_owned(),
                    };
                }
            }
            Message::Once(OnceOperation::JumpToContentView) => self.focus = Focus::ON,
            _ => {}
        }
        Message::NoAction
    }
}

impl OperatableComponent for ContentViewer {
    fn draw(&mut self, frame: &mut Frame, rect: Rect) {
        let contents: Vec<String> = self
            .content
            .lines()
            .skip(self.scroll_position)
            .take(rect.height as usize)
            .map(|line| format!("{}\n", line))
            .collect();

        let paragraph = Paragraph::new(contents.concat())
            .block(title_block(&self.title, self.focus))
            .wrap(Wrap { trim: false });

        self.context_size = Paragraph::new(self.content.clone()).line_count(rect.width);
        self.height = rect.height as usize;
        frame.render_widget(paragraph, rect)
    }

    fn process_focus(&mut self) {
        match self.focus {
            Focus::Off => self.focus = Focus::ON,
            Focus::ON => self.focus = Focus::Off,
        }
    }

    fn process_events(&mut self, events: KeyCode) -> Message {
        match events {
            KeyCode::Up => {
                if self.scroll_position > 0 {
                    self.scroll_position -= 1;
                }
            }
            KeyCode::Down => {
                // 4 is the using frame size
                if self.scroll_position < 4 + self.context_size.saturating_sub(1 + self.height) {
                    self.scroll_position += 1;
                }
            }
            KeyCode::Char('o') => {
                self.mode = self.mode.next();
                let mut repository = self.repository.lock().unwrap();
                if let Ok(rows) = repository.get_content(self.title.to_owned()) {
                    self.content = self.mode.concat(rows);
                    self.scroll_position = 0
                } else {
                    return Message::Error {
                        _message: "failed to get content".to_owned(),
                    };
                }
            }
            _ => {}
        }
        Message::NoAction
    }

    fn handle_message(&mut self, message: &Message) -> Message {
        // allow pattern
        // 1. MultipleTimes -> Once
        // 2. MultipleTimes -> NoAction
        // 3. Once -> NoAction
        // 4. NoAction -> NoAction
        match (message, self._handle_message(message)) {
            (Message::MultipleTimes(_), Message::MultipleTimes(_)) => unreachable!(),
            (Message::Once(_), Message::MultipleTimes(_)) => unreachable!(),
            (Message::Once(_), Message::Once(_)) => unreachable!(),
            (Message::NoAction, Message::MultipleTimes(_)) => unreachable!(),
            (Message::NoAction, Message::Once(_)) => unreachable!(),
            (_, new_message) => new_message,
        }
    }
}

fn title_block(title: &str, focus: Focus) -> Block {
    Block::bordered()
        .title(title.bold().into_left_aligned_line())
        .style(match focus {
            Focus::ON => Style::default(),
            Focus::Off => Style::default().fg(Color::DarkGray),
        })
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::repository::RepositoryInfo;
    use insta::assert_snapshot;
    use ratatui::{backend::TestBackend, Terminal};
    use std::sync::{Arc, Mutex};

    fn create_mock_repo() -> Arc<Mutex<RepositoryInfo>> {
        use std::env;
        use std::time::{SystemTime, UNIX_EPOCH};

        let timestamp = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_nanos();
        let random_suffix = std::process::id();
        let test_dir = env::temp_dir().join(format!(
            "gview_content_viewer_test_{}_{}",
            timestamp, random_suffix
        ));
        let _ = std::fs::remove_dir_all(&test_dir);
        std::fs::create_dir_all(&test_dir).unwrap();

        let repo = git2::Repository::init(&test_dir).unwrap();
        let signature = git2::Signature::new(
            "Test User",
            "test@example.com",
            &git2::Time::new(1234567890, 0),
        )
        .unwrap();
        let tree_id = {
            let mut index = repo.index().unwrap();
            index.write_tree().unwrap()
        };
        let tree = repo.find_tree(tree_id).unwrap();

        let _ = repo.commit(
            Some("HEAD"),
            &signature,
            &signature,
            "Initial commit",
            &tree,
            &[],
        );

        drop(tree);
        let oid = repo.head().unwrap().target().unwrap();

        let repo_info = RepositoryInfo::_from_parts(repo, oid);
        Arc::new(Mutex::new(repo_info))
    }

    #[test]
    fn test_content_viewer_draw_empty() {
        let mock_repo = create_mock_repo();
        let mut content_viewer = ContentViewer::new(mock_repo);
        content_viewer.focus = Focus::ON;
        content_viewer.title = "test.rs".to_string();
        content_viewer.content = "".to_string();

        let backend = TestBackend::new(80, 24);
        let mut terminal = Terminal::new(backend).unwrap();

        terminal
            .draw(|frame| {
                let rect = ratatui::layout::Rect::new(0, 0, 80, 24);
                content_viewer.draw(frame, rect);
            })
            .unwrap();

        let buffer = terminal.backend().buffer();
        assert_snapshot!(format!("{:?}", buffer));
    }

    #[test]
    fn test_content_viewer_draw_with_content() {
        let mock_repo = create_mock_repo();
        let mut content_viewer = ContentViewer::new(mock_repo);
        content_viewer.focus = Focus::ON;
        content_viewer.title = "main.rs".to_string();
        content_viewer.content = "fn main() {\n    println!(\"Hello, world!\");\n}\n\nfn another_function() {\n    // Some comment\n    let x = 42;\n    println!(\"x = {}\", x);\n}".to_string();

        let backend = TestBackend::new(80, 24);
        let mut terminal = Terminal::new(backend).unwrap();

        terminal
            .draw(|frame| {
                let rect = ratatui::layout::Rect::new(0, 0, 80, 24);
                content_viewer.draw(frame, rect);
            })
            .unwrap();

        let buffer = terminal.backend().buffer();
        assert_snapshot!(format!("{:?}", buffer));
    }

    #[test]
    fn test_content_viewer_draw_unfocused() {
        let mock_repo = create_mock_repo();
        let mut content_viewer = ContentViewer::new(mock_repo);
        content_viewer.focus = Focus::Off;
        content_viewer.title = "lib.rs".to_string();
        content_viewer.content = "pub fn add(left: usize, right: usize) -> usize {\n    left + right\n}\n\n#[cfg(test)]\nmod tests {\n    use super::*;\n\n    #[test]\n    fn it_works() {\n        let result = add(2, 2);\n        assert_eq!(result, 4);\n    }\n}".to_string();

        let backend = TestBackend::new(80, 24);
        let mut terminal = Terminal::new(backend).unwrap();

        terminal
            .draw(|frame| {
                let rect = ratatui::layout::Rect::new(0, 0, 80, 24);
                content_viewer.draw(frame, rect);
            })
            .unwrap();

        let buffer = terminal.backend().buffer();
        assert_snapshot!(format!("{:?}", buffer));
    }

    #[test]
    fn test_content_viewer_draw_with_line_numbers() {
        let mock_repo = create_mock_repo();
        let mut content_viewer = ContentViewer::new(mock_repo);
        content_viewer.focus = Focus::ON;
        content_viewer.title = "example.rs".to_string();
        content_viewer.mode = ShowMode::WithLine;
        content_viewer.content = "1 | use std::collections::HashMap;\n2 | \n3 | fn main() {\n4 |     let mut map = HashMap::new();\n5 |     map.insert(\"key\", \"value\");\n6 |     println!(\"{:?}\", map);\n7 | }".to_string();

        let backend = TestBackend::new(80, 24);
        let mut terminal = Terminal::new(backend).unwrap();

        terminal
            .draw(|frame| {
                let rect = ratatui::layout::Rect::new(0, 0, 80, 24);
                content_viewer.draw(frame, rect);
            })
            .unwrap();

        let buffer = terminal.backend().buffer();
        assert_snapshot!(format!("{:?}", buffer));
    }

    #[test]
    fn test_content_viewer_draw_with_blame() {
        let mock_repo = create_mock_repo();
        let mut content_viewer = ContentViewer::new(mock_repo);
        content_viewer.focus = Focus::ON;
        content_viewer.title = "blame_example.rs".to_string();
        content_viewer.mode = ShowMode::WithBlame;
        content_viewer.content = "abc123f | use std::io;\nabc123f | \n456def9 | fn main() -> Result<(), Box<dyn std::error::Error>> {\n456def9 |     let input = std::io::stdin();\n789ghi2 |     println!(\"Input received\");\n789ghi2 |     Ok(())\nabc123f | }".to_string();

        let backend = TestBackend::new(80, 24);
        let mut terminal = Terminal::new(backend).unwrap();

        terminal
            .draw(|frame| {
                let rect = ratatui::layout::Rect::new(0, 0, 80, 24);
                content_viewer.draw(frame, rect);
            })
            .unwrap();

        let buffer = terminal.backend().buffer();
        assert_snapshot!(format!("{:?}", buffer));
    }

    #[test]
    fn test_content_viewer_draw_scrolled() {
        let mock_repo = create_mock_repo();
        let mut content_viewer = ContentViewer::new(mock_repo);
        content_viewer.focus = Focus::ON;
        content_viewer.title = "scrolled.rs".to_string();
        content_viewer.scroll_position = 3;
        content_viewer.content = "Line 1\nLine 2\nLine 3\nLine 4\nLine 5\nLine 6\nLine 7\nLine 8\nLine 9\nLine 10\nLine 11\nLine 12\nLine 13\nLine 14\nLine 15\nLine 16\nLine 17\nLine 18\nLine 19\nLine 20".to_string();

        let backend = TestBackend::new(80, 24);
        let mut terminal = Terminal::new(backend).unwrap();

        terminal
            .draw(|frame| {
                let rect = ratatui::layout::Rect::new(0, 0, 80, 24);
                content_viewer.draw(frame, rect);
            })
            .unwrap();

        let buffer = terminal.backend().buffer();
        assert_snapshot!(format!("{:?}", buffer));
    }
}
