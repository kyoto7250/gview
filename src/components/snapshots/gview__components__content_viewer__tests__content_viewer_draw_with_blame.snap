---
source: src/components/content_viewer.rs
expression: "format!(\"{:?}\", buffer)"
snapshot_kind: text
---
Buffer {
    area: Rect { x: 0, y: 0, width: 80, height: 24 },
    content: [
        "┌blame_example.rs──────────────────────────────────────────────────────────────┐",
        "│abc123f | use std::io;                                                        │",
        "│abc123f |                                                                     │",
        "│456def9 | fn main() -> Result<(), Box<dyn std::error::Error>> {               │",
        "│456def9 |     let input = std::io::stdin();                                   │",
        "│789ghi2 |     println!("Input received");                                     │",
        "│789ghi2 |     Ok(())                                                          │",
        "│abc123f | }                                                                   │",
        "│                                                                              │",
        "│                                                                              │",
        "│                                                                              │",
        "│                                                                              │",
        "│                                                                              │",
        "│                                                                              │",
        "│                                                                              │",
        "│                                                                              │",
        "│                                                                              │",
        "│                                                                              │",
        "│                                                                              │",
        "│                                                                              │",
        "│                                                                              │",
        "│                                                                              │",
        "│                                                                              │",
        "└──────────────────────────────────────────────────────────────────────────────┘",
    ],
    styles: [
        x: 0, y: 0, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 0, fg: Reset, bg: Reset, underline: Reset, modifier: BOLD,
        x: 17, y: 0, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
    ]
}
